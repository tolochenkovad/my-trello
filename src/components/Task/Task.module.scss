.task {
  border: 1px solid lightgrey;
  border-radius: 2px;
  padding: 8px;
  margin-bottom: 8px;
  transition: background-color 0.2s ease;
}

.box {
  display: flex;
  flex-direction: column;
}

.body {
  display: flex;
  justify-content: space-between;
}

.content {
  width: 100%;
}

.icon {
  cursor: pointer;
  &:first-of-type {
    margin-right: 10px;
  }
}

.options {
  display: flex;
}

.icon {
  cursor: pointer;
  &:hover {
    color: #ffffff;
  }
}

.date {
  text-align: right;
  font-size: 12px;
  color: #000000;
  opacity: 0.7;
  margin-top: 5px;
}

.isDragDisabled {
  background-color: lightgrey;
}

.isDragging {
  background-color: inherit;
}

.endOfTerm {
  border: 5px solid red;
}
