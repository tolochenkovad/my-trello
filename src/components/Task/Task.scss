.task {
  border: 1px solid lightgrey;
  border-radius: 2px;
  padding: 8px;
  margin-bottom: 8px;
  transition: background-color 0.2s ease;
  background-color: #f3c197;
  &__body {
    display: flex;
    justify-content: space-between;
    .content {
      width: 100%;
    }
    .icon {
      cursor: pointer;
      &:first-of-type {
        margin-right: 10px;
      }
    }
  }
  &__options {
    display: flex;
    .icon {
      cursor: pointer;
      &:hover {
        color: #ffffff;
      }
    }
  }
  &--isDragDisabled {
    background-color: lightgrey;
  }
  &--isDragging {
    background-color: lightgreen;
  }
}