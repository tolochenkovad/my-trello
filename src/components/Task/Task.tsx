import React, { FC, useState, useMemo } from 'react';
import { Draggable } from 'react-beautiful-dnd';
import classNames from 'classnames';
import { task } from '../../types/tasks';
import { PencilSquare, Trash } from 'react-bootstrap-icons';
import { useDispatch } from 'react-redux';
import { editTask, removeTask } from '../../store/Tasks/actions';
import TaskModal from '../../common/Modals/TaskModal/TaskModal';
import ConfirmModal from '../../common/Modals/ConfirmModal/ConfirmModal';
import moment from 'moment';

type TaskType = {
  task: task;
  index: number;
};

const Task: FC<TaskType> = ({ task, index }) => {
  const [showModal, setShowModal] = useState<boolean>(false);
  const [showConfirmModal, setShowConfirmModal] = useState<boolean>(false);
  const dispatch = useDispatch();

  const editTaskContent = (value: string, color: string, dateOfTheEnd: string) => {
    setShowModal(false);
    dispatch(editTask(value, color, task.id, dateOfTheEnd));
  };

  const onConfirmModal = () => {
    setShowConfirmModal(false);
    onDeleteTask();
  };

  const onHideModal = () => {
    setShowConfirmModal(false);
  };

  const setConfirmModal = () => {
    setShowConfirmModal(true);
  };

  const onDeleteTask = () => {
    dispatch(removeTask(task.id));
  };

  const isEndOfTermTask = useMemo(() => () => {
    const currentDate: any = new Date(task.date);
    const endDate: any = new Date(task.dateOfTheEnd);
    const difference = Math.ceil((endDate - currentDate) / (60 * 60 * 24 * 1000));
    return !!(endDate && difference === 1 && task.columnId !== 'column-3');
  }, [task.date, task.dateOfTheEnd, task.columnId]);

  return (
    <>
      <Draggable draggableId={task.id} index={index}>
        {({ innerRef, draggableProps, dragHandleProps }, snapshot) => (
          <div style={{ backgroundColor: task.color }}>
            <div
              className={classNames('task', {
                'task--isDragging': snapshot.isDragging,
                'task--endOfTerm': isEndOfTermTask(),
              })}
              {...draggableProps}
              {...dragHandleProps}
              ref={innerRef}
            >
              <div className="task__box">
                <div className="task__body">
                  <div className="content">{task.content}</div>
                  <div className="task__options">
                    <div className="icon">
                      <PencilSquare onClick={() => setShowModal(true)} />
                    </div>
                    <div className="icon">
                      <Trash onClick={setConfirmModal} />
                    </div>
                  </div>
                </div>
                <div className="task__date">{moment(task.date).startOf('minutes').fromNow()}</div>
              </div>
            </div>
          </div>
        )}
      </Draggable>
      {showModal && (
        <TaskModal
          title="Edit task"
          show
          onHide={() => setShowModal(false)}
          onConfirm={editTaskContent}
          valueFromProps={task.content}
          colorFromProps={task.color}
          dateOfTheEndFromProps={task.dateOfTheEnd}
          isTheEndOfTerm={isEndOfTermTask()}
        />
      )}

      {showConfirmModal && <ConfirmModal show onHide={onHideModal} onConfirm={onConfirmModal} />}
    </>
  );
};

export default React.memo(Task);
