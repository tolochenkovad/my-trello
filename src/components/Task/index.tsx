import React, { FC, useState, useMemo } from 'react';
import { Draggable } from 'react-beautiful-dnd';
import classNames from 'classnames';
import { task } from '../../types/tasks';
import { PencilSquare, Trash } from 'react-bootstrap-icons';
import { useDispatch } from 'react-redux';
import TaskModal from '../../common/Modals/TaskModal/TaskModal';
import ConfirmModal from '../../common/Modals/ConfirmModal/ConfirmModal';
import moment from 'moment';
import { editTaskAction, removeTaskAction } from '../../store/Tasks/actions';
const styles = require('./Task.module.scss');

type TaskType = {
  task: task;
  index: number;
};

const Task: FC<TaskType> = ({ task, index }) => {
  const [showModal, setShowModal] = useState<boolean>(false);
  const [showConfirmModal, setShowConfirmModal] = useState<boolean>(false);
  const dispatch = useDispatch();

  const editTaskContent = (value: string, color: string, dateOfTheEnd: string) => {
    setShowModal(false);
    dispatch(editTaskAction.pending({ value, color, taskId: task.id, dateOfTheEnd }));
  };

  const onConfirmModal = () => {
    setShowConfirmModal(false);
    onDeleteTask();
  };

  const onHideModal = () => {
    setShowConfirmModal(false);
  };

  const setConfirmModal = () => {
    setShowConfirmModal(true);
  };

  const onDeleteTask = () => {
    dispatch(removeTaskAction.pending({ taskId: task.id }));
  };

  const isEndOfTermTask = useMemo(() => () => {
    const currentDate: any = new Date(task.date);
    const endDate: any = new Date(task.dateOfTheEnd);
    const difference = Math.ceil((endDate - currentDate) / (60 * 60 * 24 * 1000));
    return !!(endDate && difference === 1 && task.columnId !== 'column-3');
  }, [task.date, task.dateOfTheEnd, task.columnId]);

  return (
    <>
      <Draggable draggableId={task.id} index={index}>
        {({ innerRef, draggableProps, dragHandleProps }, snapshot) => (
          <div style={{ backgroundColor: task.color }}>
            <div
              className={classNames(styles.task, {
                [styles.isDragging]: snapshot.isDragging,
                [styles.endOfTerm]: isEndOfTermTask(),
              })}
              {...draggableProps}
              {...dragHandleProps}
              ref={innerRef}
            >
              <div className={styles.box}>
                <div className={styles.body}>
                  <div className={styles.content}>{task.content}</div>
                  <div className={styles.options}>
                    <div className={styles.icon}>
                      <PencilSquare onClick={() => setShowModal(true)} />
                    </div>
                    <div className={styles.icon}>
                      <Trash onClick={setConfirmModal} />
                    </div>
                  </div>
                </div>
                <div className={styles.date}>{moment(task.date).startOf('minutes').fromNow()}</div>
              </div>
            </div>
          </div>
        )}
      </Draggable>
      {showModal && (
        <TaskModal
          title="Edit task"
          show
          onHide={() => setShowModal(false)}
          onConfirm={editTaskContent}
          valueFromProps={task.content}
          colorFromProps={task.color}
          dateOfTheEndFromProps={task.dateOfTheEnd}
          isTheEndOfTerm={isEndOfTermTask()}
        />
      )}

      {showConfirmModal && <ConfirmModal show onHide={onHideModal} onConfirm={onConfirmModal} />}
    </>
  );
};

export default React.memo(Task);
