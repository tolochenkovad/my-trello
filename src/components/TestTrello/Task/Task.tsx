import React from 'react';
import { Draggable } from 'react-beautiful-dnd';
import { task } from '../initial-data';
import classNames from 'classnames';

type TaskType = {
  task: task;
  index: number;
};

const Task: React.FC<TaskType> = ({ task, index }) => (
  <Draggable draggableId={task.id} index={index}>
    {({ innerRef, draggableProps, dragHandleProps }, snapshot) => (
      <div
        className={classNames('task', {
          'task--isDragging': snapshot.isDragging,
        })}
        {...draggableProps}
        {...dragHandleProps}
        ref={innerRef}
      >
        {task.content}
      </div>
    )}
  </Draggable>
);

export default Task;
