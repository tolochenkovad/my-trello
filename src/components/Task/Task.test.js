import React from 'react';
import { shallow } from 'enzyme';
import Task from './Task';

const props = {
  task: { id: '1', content: 'test', columnId: '1', date: 'date', dateOfTheEnd: 'dateOfTheEnd', color: '#000000' },
  index: 1,
};

let wrapper;

beforeEach(() => {
  wrapper = shallow(<Task {...props} />);
});

describe('Task', () => {
  it('should render', () => {
    expect(wrapper).toHaveLength(1);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
