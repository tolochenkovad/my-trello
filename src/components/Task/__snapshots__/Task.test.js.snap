// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task should render correctly 1`] = `
<Fragment>
  <Draggable
    draggableId="1"
    index={1}
  >
    <Component />
  </Draggable>
</Fragment>
`;
